### PR TITLE
Handle type mismatch for complex array initialization

### PR DESCRIPTION
**Pull request by Kishan-Ved**
Fixes: https://github.com/lfortran/lfortran/issues/3317

### CODE REVIEW

**Comment by Kishan-Ved** [comment]
@HarshitaKalani This builds without errors but doesn't fix the issue. What else needs to be changed?

**Comment by Kishan-Ved** [comment]
Also, E(3) = [(1, 1), (2, 2), (3, 3)] does this mean 1 + 1*j, 2 + 2*j and 3 + 3*j or 6 different complex numbers with imaginary part = 0?

**Comment by HarshitaKalani** [comment]
>This builds without errors but doesn't fix the issue. 

This fixes another issue when we have `complex(8) E(3) = [1, 2, 3]` since you did it for IntegerToComplex.

However, here you need to deal with ComplexToComplex as:
complex(8) E(3) = [(1, 1), (2, 2), (3, 3)] means 1 + 1j, 2 + 2j and 3 + 3j, where (1, 1) is complex(4). Convert this to complex(8) by converting (1, 1) to (1.0000000000000000,1.0000000000000000)

You can also do it for ComplexToReal, RealToComplex as well as ComplexToInteger and add tests to check them all.

**Comment by Kishan-Ved** [comment]
> > This builds without errors but doesn't fix the issue.
> 
> This fixes another issue when we have `complex(8) E(3) = [1, 2, 3]` since you did it for IntegerToComplex.
> 
> However, here you need to deal with ComplexToComplex as: complex(8) E(3) = [(1, 1), (2, 2), (3, 3)] means 1 + 1j, 2 + 2j and 3 + 3j, where (1, 1) is complex(4). Convert this to complex(8) by converting (1, 1) to (1.0000000000000000,1.0000000000000000)
> 
> You can also do it for ComplexToReal, RealToComplex as well as ComplexToInteger and add tests to check them all.
Thanks, I'll work on this tomorrow.

**Comment by Kishan-Ved** [comment]
> > This builds without errors but doesn't fix the issue.
> 
> This fixes another issue when we have `complex(8) E(3) = [1, 2, 3]` since you did it for IntegerToComplex.
> 
> However, here you need to deal with ComplexToComplex as: complex(8) E(3) = [(1, 1), (2, 2), (3, 3)] means 1 + 1j, 2 + 2j and 3 + 3j, where (1, 1) is complex(4). Convert this to complex(8) by converting (1, 1) to (1.0000000000000000,1.0000000000000000)
> 
> You can also do it for ComplexToReal, RealToComplex as well as ComplexToInteger and add tests to check them all.

Yes, this works for:
```
program main
    complex(8) :: E(3) = [1, 2, 3]
    print*, E
end
```
Output:
```
(lf) kishan@kishan-IdeaPad-3-15IIL05:~/Desktop/lfortran$ ./inst/bin/lfortran kishan.f90
(1.000000,0.000000) (2.000000,0.000000) (3.000000,0.000000) 
```

**Comment by certik** [comment]
This PR needs tests.

**Comment by Kishan-Ved** [comment]
types_26.f90 fails to build and I am getting this error. Please help.
```
  File "/home/kishan/Desktop/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 2503, in LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_DeclarationUtil(LCompilers::LFortran::AST::Declaration_t const&)
    ImplicitCastRules::set_converted_value(al, x.base.base.loc, &init_expr, init_type, type);
AssertFailed: ASR::is_a<ASR::Real_t>(*ASRUtils::type_get_past_pointer(dest_type))
make[2]: *** [CMakeFiles/types_26.dir/build.make:75: CMakeFiles/types_26.dir/types_26.f90.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:22315: CMakeFiles/types_26.dir/all] Error 2
make: *** [Makefile:101: all] Error 2
Command failed.
```

**Comment by Kishan-Ved** [comment]
I've currently commented out types_24 and types_26 from the CMakeList, this is because I get the error: 
```
Traceback (most recent call last):
  Binary file "/home/kishan/Desktop/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main_impl()
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main()
  File "/home/kishan/Desktop/lfortran/src/bin/lfortran.cpp", line 2375, in ??
    return main_app(argc, argv);
  File "/home/kishan/Desktop/lfortran/src/bin/lfortran.cpp", line 2304, in main_app(int, char**)
    return compile_to_object_file(arg_file, outfile, false,
  File "/home/kishan/Desktop/lfortran/src/bin/lfortran.cpp", line 907, in ??
    result = fe.get_asr2(input, lm, diagnostics);
  File "/home/kishan/Desktop/lfortran/src/lfortran/fortran_evaluator.cpp", line 253, in LCompilers::FortranEvaluator::get_asr2(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::LocationManager&, LCompilers::diag::Diagnostics&)
    Result<ASR::TranslationUnit_t*> res2 = get_asr3(*ast, diagnostics);
  File "/home/kishan/Desktop/lfortran/src/lfortran/fortran_evaluator.cpp", line 275, in LCompilers::FortranEvaluator::get_asr3(LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&)
    compiler_options.symtab_only, compiler_options);
  File "/home/kishan/Desktop/lfortran/src/lfortran/semantics/ast_to_asr.cpp", line 97, in LCompilers::LFortran::ast_to_asr(Allocator&, LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&, LCompilers::SymbolTable*, bool, LCompilers::CompilerOptions&)
    instantiate_types, instantiate_symbols, entry_functions, entry_function_arguments_mapping, data_structure);
  File "/home/kishan/Desktop/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 3460, in LCompilers::LFortran::symbol_table_visitor(Allocator&, LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&, LCompilers::SymbolTable*, LCompilers::CompilerOptions&, std::map<unsigned long, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > > > > >&, std::map<unsigned long, LCompilers::ASR::symbol_t*, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, LCompilers::ASR::symbol_t*> > >&, std::map<unsigned long, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >&, std::map<unsigned int, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > > > > >&, std::map<unsigned int, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::symbol_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::symbol_t*> > >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::symbol_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::symbol_t*> > > > > >&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> > > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> > > > > > > >&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<int, std::allocator<int> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<int, std::allocator<int> > > > >&, std::vector<LCompilers::ASR::stmt_t*, std::allocator<LCompilers::ASR::stmt_t*> >&)
    v.visit_TranslationUnit(ast);
  File "/home/kishan/Desktop/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 120, in LCompilers::LFortran::SymbolTableVisitor::visit_TranslationUnit(LCompilers::LFortran::AST::TranslationUnit_t const&)
    visit_ast(*x.m_items[i]);
  File "/home/kishan/Desktop/lfortran/src/lfortran/ast.h", line 4848, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_ast(LCompilers::LFortran::AST::ast_t const&)
    void visit_ast(const ast_t &b) { visit_ast_t(b, self()); }
  File "/home/kishan/Desktop/lfortran/src/lfortran/ast.h", line 4805, in ??
    case astType::mod: { v.visit_mod((const mod_t &)x); return; }
  File "/home/kishan/Desktop/lfortran/src/lfortran/ast.h", line 4851, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_mod(LCompilers::LFortran::AST::mod_t const&)
    void visit_mod(const mod_t &b) { visit_mod_t(b, self()); }
  File "/home/kishan/Desktop/lfortran/src/lfortran/ast.h", line 4439, in ??
    case modType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/home/kishan/Desktop/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 521, in LCompilers::LFortran::SymbolTableVisitor::visit_Program(LCompilers::LFortran::AST::Program_t const&)
    visit_unit_decl2(*x.m_decl[i]);
  File "/home/kishan/Desktop/lfortran/src/lfortran/ast.h", line 4862, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_unit_decl2(LCompilers::LFortran::AST::unit_decl2_t const&)
    void visit_unit_decl2(const unit_decl2_t &b) { visit_unit_decl2_t(b, self()); }
  File "/home/kishan/Desktop/lfortran/src/lfortran/ast.h", line 4465, in ??
    case unit_decl2Type::Declaration: { v.visit_Declaration((const Declaration_t &)x); return; }
  File "/home/kishan/Desktop/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 1555, in LCompilers::LFortran::SymbolTableVisitor::visit_Declaration(LCompilers::LFortran::AST::Declaration_t const&)
    visit_DeclarationUtil(x);
  File "/home/kishan/Desktop/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 2503, in LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_DeclarationUtil(LCompilers::LFortran::AST::Declaration_t const&)
    ImplicitCastRules::set_converted_value(al, x.base.base.loc, &init_expr, init_type, type);
AssertFailed: ASR::is_a<ASR::Integer_t>(*ASRUtils::type_get_past_pointer(dest_type))
make[2]: *** [CMakeFiles/types_24.dir/build.make:75: CMakeFiles/types_24.dir/types_24.f90.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:22261: CMakeFiles/types_24.dir/all] Error 2
make: *** [Makefile:101: all] Error 2
Command failed.
```
So ComplextoReal and ComplextoInteger are causing some assertion issue.

**Review by HarshitaKalani — COMMENTED** [review]
In the `src/lfortran/semantics/asr_implicit_cast_rules.h` file, apply the following diff:
```diff
         } else if ((ASR::cast_kindType)cast_kind == ASR::cast_kindType::ComplexToInteger) {
             if (ASRUtils::expr_value(*convert_can)) {
-                LCOMPILERS_ASSERT(ASR::is_a<ASR::Integer_t>(*ASRUtils::type_get_past_pointer(dest_type)))
+                LCOMPILERS_ASSERT(ASR::is_a<ASR::Integer_t>(*ASRUtils::type_get_past_pointer(dest_type2)))
                 LCOMPILERS_ASSERT(ASR::is_a<ASR::Complex_t>(*ASRUtils::type_get_past_array(ASRUtils::expr_type(*convert_can))))
                 value = ASRUtils::expr_value(*convert_can);
-                LCOMPILERS_ASSERT(ASR::is_a<ASR::ComplexConstant_t>(*value))
-                ASR::ComplexConstant_t *r = ASR::down_cast<ASR::ComplexConstant_t>(value);
-                double rval = r->m_re;
-                int64_t ival = (int64_t)rval;
-                value = (ASR::expr_t *)ASR::make_IntegerConstant_t(al, a_loc,
-                    ival, dest_type2);
+                if (ASR::is_a<ASR::ComplexConstant_t>(*value)) {
+                  ASR::ComplexConstant_t *r = ASR::down_cast<ASR::ComplexConstant_t>(value);
+                  double rval = r->m_re;
+                  int64_t ival = (int64_t)rval;
+                  value = (ASR::expr_t *)ASR::make_IntegerConstant_t(al, a_loc,
+                      ival, dest_type2);
+                } else {
+                  value = nullptr;
+                }
```
Ref: https://github.com/lfortran/lfortran/pull/3316/files#r1483090123

File: integration_tests/CMakeLists.txt [hunk 1]
```diff
@@ -1376,3 +1376,7 @@ RUN(NAME elemental_08 LABELS gfortran llvm)
 
 RUN(NAME types_21 LABELS gfortran llvm)
 RUN(NAME types_22 LABELS gfortran llvm)
+RUN(NAME types_23 LABELS gfortran llvm)
+# RUN(NAME types_24 LABELS gfortran llvm)
+RUN(NAME types_25 LABELS gfortran llvm)
+# RUN(NAME types_26 LABELS gfortran llvm)
```
Comment (HarshitaKalani) [inline integration_tests/CMakeLists.txt:1382]: Also, you need not write 4 different tests, you can just add them all in one test.

Comment (Kishan-Ved) [reply integration_tests/CMakeLists.txt]: Done

**Comment by Kishan-Ved** [comment]
> In the `src/lfortran/semantics/asr_implicit_cast_rules.h` file, apply the following diff:
> 
> ```diff
>          } else if ((ASR::cast_kindType)cast_kind == ASR::cast_kindType::ComplexToInteger) {
>              if (ASRUtils::expr_value(*convert_can)) {
> -                LCOMPILERS_ASSERT(ASR::is_a<ASR::Integer_t>(*ASRUtils::type_get_past_pointer(dest_type)))
> +                LCOMPILERS_ASSERT(ASR::is_a<ASR::Integer_t>(*ASRUtils::type_get_past_pointer(dest_type2)))
>                  LCOMPILERS_ASSERT(ASR::is_a<ASR::Complex_t>(*ASRUtils::type_get_past_array(ASRUtils::expr_type(*convert_can))))
>                  value = ASRUtils::expr_value(*convert_can);
> -                LCOMPILERS_ASSERT(ASR::is_a<ASR::ComplexConstant_t>(*value))
> -                ASR::ComplexConstant_t *r = ASR::down_cast<ASR::ComplexConstant_t>(value);
> -                double rval = r->m_re;
> -                int64_t ival = (int64_t)rval;
> -                value = (ASR::expr_t *)ASR::make_IntegerConstant_t(al, a_loc,
> -                    ival, dest_type2);
> +                if (ASR::is_a<ASR::ComplexConstant_t>(*value)) {
> +                  ASR::ComplexConstant_t *r = ASR::down_cast<ASR::ComplexConstant_t>(value);
> +                  double rval = r->m_re;
> +                  int64_t ival = (int64_t)rval;
> +                  value = (ASR::expr_t *)ASR::make_IntegerConstant_t(al, a_loc,
> +                      ival, dest_type2);
> +                } else {
> +                  value = nullptr;
> +                }
> ```
> 
> Ref: https://github.com/lfortran/lfortran/pull/3316/files#r1483090123
> 
> This is for ComplexToInteger, follow the same for others.

Thanks, I've done this for the needed  cases.

**Review by HarshitaKalani — COMMENTED** [review]
Looks good to me!
Just check this case more. If this does not work yet, just open a new issue (RealToComplex) and mark it as `easy to fix`.
```
complex(8) :: D(3) = [1.0, 2.0, 3.0]
    print*, D
```

**Comment by Kishan-Ved** [comment]
> Looks good to me! Just check this case more. If this does not work yet, just open a new issue (RealToComplex) and mark it as `easy to fix`.
> 
> ```
> complex(8) :: D(3) = [1.0, 2.0, 3.0]
>     print*, D
> ```

Created an issue here: https://github.com/lfortran/lfortran/issues/3372. I don't have the permissions to assign labels.

**Review by certik — APPROVED** [review]
I think that's fine, thanks!

**Pull request merged**